File: iosApp/iosApp/data/mappers/chats/ChatsMapper.swift
//
//  ChatsMapper.swift
//  iosApp
//
//  Created by nVoxel on 25.04.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import TDLibKit
import TGDriveKit

class ChatsMapper {
    
    private let fileMapper: FileMapper
    
    init(fileMapper: FileMapper) {
        self.fileMapper = fileMapper
    }
    
    func toResponse(chat: Chat) -> CommonChat {
        return CommonChat(
            id: chat.id,
            type: toResponse(chatType: chat.type),
            title: chat.title,
            photo: chat.photo != nil ? toResponse(chatPhotoInfo: chat.photo!) : nil,
            hasProtectedContent: chat.hasProtectedContent
        )
    }
    
    private func toResponse(chatType: ChatType) -> CommonChatType {
        switch chatType {
        case .chatTypePrivate(let privateChat):
            return CommonChatTypePrivate(userId: privateChat.userId)
        case .chatTypeBasicGroup(let basisGroupChat):
            return CommonChatTypeBasicGroup(basicGroupId: basisGroupChat.basicGroupId)
        case .chatTypeSupergroup(let supergroupChat):
            return CommonChatTypeSupergroup(supergroupId: supergroupChat.supergroupId, isChannel: supergroupChat.isChannel)
        case .chatTypeSecret(let secretChat):
            return CommonChatTypeSecret(secretChatId: Int64(secretChat.secretChatId), userId: secretChat.userId)
        }
    }
    
    private func toResponse(chatPhotoInfo: ChatPhotoInfo) -> CommonChatPhotoInfo {
        return CommonChatPhotoInfo(
            small: fileMapper.toResponse(file: chatPhotoInfo.small),
            big: fileMapper.toResponse(file: chatPhotoInfo.big)
        )
    }
}

File: source/data/mappers/src/jvmMain/kotlin/com/voxeldev/tgdrive/mappers/chats/ChatsMapper.kt
package com.voxeldev.tgdrive.mappers.chats

import com.voxeldev.tgdrive.mappers.media.FileMapper
import org.drinkless.tdlib.TdApi.Chat
import org.drinkless.tdlib.TdApi.ChatPhotoInfo
import org.drinkless.tdlib.TdApi.ChatType
import org.drinkless.tdlib.TdApi.ChatTypeBasicGroup
import org.drinkless.tdlib.TdApi.ChatTypePrivate
import org.drinkless.tdlib.TdApi.ChatTypeSecret
import org.drinkless.tdlib.TdApi.ChatTypeSupergroup
import com.voxeldev.tgdrive.main.api.chats.Chat as CommonChat
import com.voxeldev.tgdrive.main.api.chats.ChatPhotoInfo as CommonChatPhotoInfo
import com.voxeldev.tgdrive.main.api.chats.ChatType as CommonChatType

/**
 * @author nvoxel
 */
class ChatsMapper(
    private val fileMapper: FileMapper,
) {

    fun toResponse(chat: Chat): CommonChat =
        CommonChat(
            id = chat.id,
            type = chat.type.toResponse(),
            title = chat.title,
            photo = chat.photo?.let { chat.photo.toResponse() },
            hasProtectedContent = chat.hasProtectedContent,
        )

    private fun ChatType.toResponse(): CommonChatType? =
        when (this) {
            is ChatTypePrivate -> CommonChatType.Private(userId = userId)
            is ChatTypeBasicGroup -> CommonChatType.BasicGroup(basicGroupId = basicGroupId)
            is ChatTypeSupergroup -> CommonChatType.Supergroup(supergroupId = supergroupId, isChannel = isChannel)
            is ChatTypeSecret -> CommonChatType.Secret(secretChatId = secretChatId.toLong(), userId = userId)
            else -> null
        }

    private fun ChatPhotoInfo.toResponse(): CommonChatPhotoInfo =
        CommonChatPhotoInfo(
            small = fileMapper.toResponse(small),
            big = fileMapper.toResponse(big),
        )
}
